### PR TITLE
Initial version of SPD aggregate

### PR DESCRIPTION
**Pull request by Bradeskojest**

### CODE REVIEW

File: src/third_party/geospatial/geospatial_aggr.cpp [hunk 1]
```diff
@@ -43,8 +174,180 @@ TStayPointDetector::TStayPointDetector(const TWPt<TBase>& Base, const PJsonVal&
...
+        gps.Distance = 0;
+    }
+    return true;
+}//parseRecord
+
+    /// Helper function for easier GPSRecord creation. It assigns values to new
+    /// TGPSRecord, compares it to the previous record and then re-stores it as
+    /// lastRecord, for next parsing
+TGPSMeasurement* TStayPointDetector::PrepareGPSRecord(const TRec& Rec) {
+    TGPSMeasurement gps;
```
Comment (rupnikj) [inline src/third_party/geospatial/geospatial_aggr.cpp]: simplify using queue

File: src/third_party/geospatial/geospatial_aggr.h [hunk 2]
```diff
@@ -16,10 +17,98 @@
 #define __STDC_FORMAT_MACROS
 #include <inttypes.h>
 
+#include <iostream>
```
Comment (rupnikj) [inline src/third_party/geospatial/geospatial_aggr.h]: remove STL stuff and #include <inttypes.h> - to print uint64 use TUInt64::GetStr()

File: src/third_party/geospatial/geospatial_aggr.h [hunk 3]
```diff
@@ -16,10 +17,98 @@
...
 namespace TQm {
 namespace TStreamAggrs {
 
+//////////////////////
+/// representation of one GPS point
+struct TPoint {
+    TFlt Lat;
+    TFlt Lon;
+    TPoint() : Lat(0), Lon(0) {}
+    TPoint(TFlt lat, TFlt lon) : Lat(lat), Lon(lon) {}
```
Comment (rupnikj) [inline src/third_party/geospatial/geospatial_aggr.h]: caps

File: src/third_party/geospatial/geospatial_aggr.h [hunk 4]
```diff
@@ -16,10 +17,98 @@
...
+    TGPSMeasurement();
+    TPoint LatLon;
+    TUInt64 Time = 0;//timestamp
+    TFlt Accuracy = 0;//accuracy
+    TFlt Speed = -1;//given speed by GPS
+    TFlt Distance = -1;//distance to previous
+    TUInt64 TimeDiff = 0;//time difference with previous
+
+    void PrintConsole() {
+        std::cout << "GPS:{time:" << Time << "\n";
```
Comment (rupnikj) [inline src/third_party/geospatial/geospatial_aggr.h]: printf("%s", Str.CStr())

File: src/third_party/geospatial/geospatial_aggr.h [hunk 5]
```diff
@@ -16,10 +17,98 @@
...
+
+    PJsonVal ToJson() const;
+};
+
+//////////////////////
+/// provides common Geo utilities
+class TGeoUtils {
+private:
+    const static int R = 6371000;//earth radius in m
+    const static double Pi;//Pi
```
Comment (rupnikj) [inline src/third_party/geospatial/geospatial_aggr.h]: TMath::Pi

File: src/third_party/geospatial/geospatial_aggr.h [hunk 6]
```diff
@@ -16,10 +17,98 @@
...
+public:
+    /// Calculates distance betweet 2 geo points based on
+    /// equirectangular distance approximation - this is
+    /// good enough for small distances
+    static double QuickDist(const TPoint& p1, const TPoint& p2);
```
Comment (rupnikj) [inline src/third_party/geospatial/geospatial_aggr.h]: caps

File: src/third_party/geospatial/geospatial_aggr.h [hunk 7]
```diff
@@ -16,10 +17,98 @@
...
+enum TGeoActivityType { Staytpoint, Path };
+enum TGeoActivityStatus { Current, Possible, Detected };
+class TGeoCluster {
+private:
+    TInt MStartIdx = -1;//start pointer to GPS Measurements state array
+    TInt MEndIdx = -1;  //end pointer to GPS Measurements state array
+    TUInt64 Arrive = 0;
+    TUInt64 Depart = 0;
+    TPoint CenterPoint = TPoint(0, 0);
+	TFlt AvgSpeed = 0;
```
Comment (rupnikj) [inline src/third_party/geospatial/geospatial_aggr.h]: indent

File: src/third_party/geospatial/geospatial_aggr.h [hunk 8]
```diff
@@ -16,10 +17,98 @@
...
+	TFlt AvgAccuracy = 0;
+	TFlt Distance = 0;
+    TGeoActivityType GeoType = TGeoActivityType::Path;
+    TGeoActivityStatus GeoActStatus = TGeoActivityStatus::Current;
+public:
+    TGeoCluster(const int StartIdx, const int EndIdx,
```
Comment (rupnikj) [inline src/third_party/geospatial/geospatial_aggr.h]: const int&

File: src/third_party/geospatial/geospatial_aggr.h [hunk 9]
```diff
@@ -16,10 +17,98 @@
...
+    TGeoCluster() : GeoType(TGeoActivityType::Path) {};
+    void AddPoint(const TInt Idx, const TVec<TGPSMeasurement>& _GpsState);
+    uint Duration();
+    uint Len() const;
+    uint EndIdx() const { return MEndIdx; }
+    uint StartIdx() const { return MStartIdx; }
+    void DownShiftIdx(int _Num) { MStartIdx -= _Num; MEndIdx -= _Num; }
+    const TPoint& Center() const { return CenterPoint; }
+    const TGeoActivityType Type() const { return GeoType; }
+    void SetStatus(TGeoActivityStatus _status) {
```
Comment (rupnikj) [inline src/third_party/geospatial/geospatial_aggr.h]: caps

File: src/third_party/geospatial/geospatial_aggr.h [hunk 10]
```diff
@@ -34,8 +123,26 @@ class TStayPointDetector : public TStreamAggr {
     TInt AccuracyFieldId;
     /// time field id for fast access
     TInt TimeFieldId;
-    // algorithm parameters
+
+    /// algorithm parameters
     PJsonVal Params;
+    /// distance threshold (meters)
+    TUInt TrDist = 50;
```
Comment (rupnikj) [inline src/third_party/geospatial/geospatial_aggr.h]: constructor with default parameters

File: src/third_party/geospatial/geospatial_aggr.h [hunk 11]
```diff
@@ -34,8 +123,26 @@ class TStayPointDetector : public TStreamAggr {
...
+    /// time threshold (seconds)
+    TUInt TrTime = 300;
+
+    /// state
+    //list of GPS measurements currently part of the satate
+    TVec<TGPSMeasurement> StateGpsMeasurements;
+    //TGPSMeasurement lastRecord;//holds previous GPS record
+    TGeoCluster cl;//cluster from the algorihm (cl)
```
Comment (rupnikj) [inline src/third_party/geospatial/geospatial_aggr.h]: caps

File: src/third_party/geospatial/geospatial_aggr.h [hunk 12]
```diff
@@ -34,8 +123,26 @@ class TStayPointDetector : public TStreamAggr {
...
+    TGeoCluster plocs;//path locs from the algorithm (plocs)
+    TVec<TGeoCluster> DetectedGeoActivities;
+    bool hasFinishedGeoActs = false;
+
+    /// helper methods
+    bool parseRecord(const TRec& Rec, TGPSMeasurement& gps);
```
Comment (rupnikj) [inline src/third_party/geospatial/geospatial_aggr.h]: caps

File: src/third_party/geospatial/geospatial_aggr.cpp [hunk 13]
```diff
@@ -11,30 +13,159 @@
...
+	AvgAccuracy = AvgAccuracy +
+		((CurrentGPS.Accuracy - AvgAccuracy) / Len);
+	
+	//distance
+	Distance = Distance + CurrentGPS.Distance;
+}//TGeoCluster::addPoint
+
+/// returns duration in seconds
+uint TGeoCluster::Duration() {
+    return (int)(Depart - Arrive) / 1000;
```
Comment (blazf) [inline src/third_party/geospatial/geospatial_aggr.cpp]: cast (`int`) should match return type (`uint`)

File: src/third_party/geospatial/geospatial_aggr.cpp [hunk 14]
```diff
@@ -43,8 +174,180 @@ TStayPointDetector::TStayPointDetector(const TWPt<TBase>& Base, const PJsonVal&
...
+void TStayPointDetector::OnAddRec(const TRec& Rec,
+    const TWPt<TStreamAggr>& CallerAggr)
+{
+    TScopeStopWatch StopWatch(ExeTm);//auto stopwatch for the QM aggregate
+                                        //parse new GPS record from JSON
+
+    int NumOfRecordsToForget = 0;
+    hasFinishedGeoActs = false;
+    //delete previously detected stuff
+    for (int i = 0; i < DetectedGeoActivities.Len(); i++) {
```
Comment (blazf) [inline src/third_party/geospatial/geospatial_aggr.cpp]: Typically we use suffix V to indicate vectors, e.g. `DetectedGeoActivities` => `DetectedGeoActivityV`

File: src/third_party/geospatial/geospatial_aggr.h [hunk 15]
```diff
@@ -16,10 +17,98 @@
...
+
+//////////////////////
+/// representation of one GPS measurement
+class TGPSMeasurement {
+private:
+    //PJsonVal Json;
+public:
+    TGPSMeasurement();
+    TPoint LatLon;
+    TUInt64 Time = 0;//timestamp
```
Comment (blazf) [inline src/third_party/geospatial/geospatial_aggr.h]: Default values should be specified in constructors.

File: src/third_party/geospatial/geospatial_aggr.cpp [hunk 16]
```diff
@@ -11,30 +13,159 @@
...
-    State->AddToObj("parameters", Params);
-    // copy record
-    TTm Timestamp; Rec.GetFieldTm(TimeFieldId, Timestamp);
-    State->AddToObj("lastTimestamp",
-        (double)TTm::GetUnixMSecsFromWinMSecs(Timestamp.GetMSecsFromTm(Timestamp)));
-    State->AddToObj("lastLocationX", Rec.GetFieldFltPr(LocationFieldId).Val1);
-    State->AddToObj("lastLocationY", Rec.GetFieldFltPr(LocationFieldId).Val2);
-    State->AddToObj("lastAccuracy", Rec.GetFieldByte(AccuracyFieldId));
+/// TStaypointCluster
+void TGeoCluster::AddPoint(const TInt Idx,
```
Comment (Bradeskojest) [inline src/third_party/geospatial/geospatial_aggr.cpp]: Changing this. I did a very quick search and it's not clear to me (and some other people) whether passing int as a reference makes the code slower: http://stackoverflow.com/questions/7701221/how-to-pass-literal-integers-by-reference-in-c-newbie

File: src/third_party/geospatial/geospatial_aggr.h [hunk 17]
```diff
@@ -16,10 +17,85 @@
...
+private:
+	TInt MStartIdx;//start pointer to GPS Measurements state array
+	TInt MEndIdx;  //end pointer to GPS Measurements state array
+	TUInt64 Arrive;
+	TUInt64 Depart;
+	TPoint CenterPoint;
+	TFlt AvgSpeed;
+	TFlt AvgAccuracy;
+	TFlt Distance;
+	TGeoActivityType GeoType = TGeoActivityType::Path;
```
Comment (rupnikj) [inline src/third_party/geospatial/geospatial_aggr.h]: move to constructor

**Pull request merged**